hbase-handler/src/java/org/apache/hadoop/hive/hbase/HiveHBaseTableInputFormatV2.java

```diff
@@ -0,0 +1,375 @@
+
+  static final Logger LOG = LoggerFactory.getLogger(HiveHBaseTableInputFormatV2.class);
+  private static final Object HBASE_TABLE_MONITOR = new Object();
+
+  private HiveHBaseTableInputFormatDelegate delegate = new HiveHBaseTableInputFormatDelegate();
+
+  @Override public RecordReader<ImmutableBytesWritable, ResultWritable> getRecordReader(InputSplit split,
+                                                                                        JobConf jobConf, final Reporter reporter) throws IOException {
+
+    HBaseSplit hbaseSplit = (HBaseSplit) split;
+    TableSplit tableSplit = hbaseSplit.getTableSplit();
+
+    final org.apache.hadoop.mapreduce.RecordReader<ImmutableBytesWritable, Result> recordReader;
+
+    Job job = new Job(jobConf);
+    TaskAttemptContext tac = ShimLoader.getHadoopShims().newTaskAttemptContext(job.getConfiguration(), reporter);
+
+    final Connection conn;
+
+    synchronized (HBASE_TABLE_MONITOR) {
+      conn = ConnectionFactory.createConnection(HBaseConfiguration.create(jobConf));
+      delegate.initializeTableDelegate(conn, tableSplit.getTable());
+      delegate.setScan(HiveHBaseInputFormatUtil.getScan(jobConf));
+      recordReader = delegate.createRecordReader(tableSplit, tac);
+      try {
+        recordReader.initialize(tableSplit, tac);
+      } catch (InterruptedException e) {
+        delegate.closeTableDelegate(); // Free up the HTable connections
+        conn.close();
+        throw new IOException("Failed to initialize RecordReader: ", e);
+      }
+    }
+
+    return new RecordReader<ImmutableBytesWritable, ResultWritable>() {
+
+      @Override public void close() throws IOException {
+        synchronized (HBASE_TABLE_MONITOR) {
+          recordReader.close();
+          delegate.closeTableDelegate();
+          conn.close();
+        }
+      }
+
+      @Override public ImmutableBytesWritable createKey() {
+        return new ImmutableBytesWritable();
+      }
+
+      @Override public ResultWritable createValue() {
+        return new ResultWritable(new Result());
+      }
+
+      @Override public long getPos() throws IOException {
+        return 0;
+      }
+
+      @Override public float getProgress() throws IOException {
+        float progress = 0.0F;
+
+        try {
+          progress = recordReader.getProgress();
+        } catch (InterruptedException e) {
+          throw new IOException(e);
+        }
+
+        return progress;
+      }
+
+      @Override public boolean next(ImmutableBytesWritable rowKey, ResultWritable value) throws IOException {
+
+        boolean next = false;
+
+        try {
+          next = recordReader.nextKeyValue();
+
+          if (next) {
+            rowKey.set(recordReader.getCurrentValue().getRow());
+            value.setResult(recordReader.getCurrentValue());
+          }
+        } catch (InterruptedException e) {
+          throw new IOException(e);
+        }
+
+        return next;
+      }
+    };
+  }
```

Comment:
this is an exact dupe from ``HiveHBaseTableInputFormat``, avoid duplication and refactor